src/Utilities/hooks/useFetchOperatingSystems.js

```diff
@@ -1,32 +1,45 @@
 import { useCallback, useEffect, useRef, useState } from 'react';
 import { getOperatingSystems } from '../../api';
 
+const initialState = {
+  operatingSystems: [],
+  operatingSystemsLoaded: false,
+};
+
 const useFetchOperatingSystems = ({
   apiParams = [],
   hasAccess,
   showCentosVersions,
+  fetchCustomOSes,
 }) => {
   const mounted = useRef(true);
-  const initialState = {
-    operatingSystems: [],
-    operatingSystemsLoaded: false,
-  };
+
   const [data, setData] = useState(initialState);
 
   const fetchOperatingSystems = useCallback(async () => {
-    if (typeof hasAccess === 'undefined') {
-      return await getOperatingSystems(apiParams, showCentosVersions);
-    } else if (typeof hasAccess !== 'undefined' && hasAccess === true) {
-      return await getOperatingSystems(apiParams, showCentosVersions);
+    if (hasAccess === false) return;
+
+    const fetchFn = fetchCustomOSes || getOperatingSystems;
+    const fetchArgs = [apiParams, showCentosVersions];
+
+    try {
+      return await fetchFn(...fetchArgs);
+    } catch (error) {
+      return { results: [] };
     }
-  }, [hasAccess, apiParams, showCentosVersions]);
+  }, [
+    hasAccess,
+    JSON.stringify(apiParams),
+    showCentosVersions,
+    typeof fetchCustomOSes !== 'undefined',
+  ]);
 
   useEffect(() => {
     (async () => {
       const result = await fetchOperatingSystems();
       if (mounted.current) {
         setData({
-          operatingSystems: result.results,
+          operatingSystems: result?.results || [],
```

Comment:
Yes, such a function defiantly needed. However, each consumer may have different API response shape. Thus, I think it is better to put that function in those apps. The consumer API fetch function should provide already suitable data to inventory os filter. Have a look at compliance: https://github.com/RedHatInsights/compliance-frontend/pull/2192/files#diff-e0494b62ef3cc7ba3308fa54b1933e5d3c299e030a4ac5adaef91ef23630a08dR31

Reply:
I will update the readme with this information

Reply:
> However, each consumer may have different API response shape.

This is exactly what I wanted to say above. This function is a must IMHO so that whenever we use InventoryTable we can pass, let's call it, `processCustomOsResponse`.

 ```suggestion
          operatingSystems: processCustomOsResponse ? processCustomOsResponse(result) || results?.results || [],
```

Reply:
Would we create another prop to pass through the `processCustomOsResponse`? Would not it result in more props than needed?

Or are you suggesting to write the function inside inventory itself? If yes, I am afraid this function would not make sense as it would not be 'aware' of all possible API result shapes from custom endpoints.

I believe, the cleanest is to leave the everything related to fetching the os versions and making the their result compatible with operating system hook requirements.

Reply:
> Would not it result in more props than needed?
I am not sure I understand.

> making the their result compatible with operating system hook requirements
Which basically translates to "make all custom APIs to be adjusted to this hook expectations," which is not preferable requirement. It is way easier to add a custom function now rather than altering API responses.

That's up on you. I have unresolved this thread because I feel it's needed. If it's not a use case now, then we can deal with it later.

Reply:
> I am not sure I understand.

By more props, I mean we would need to pass not only a prop for the custom operating systems fetching, but also we would then need to pass another prop to process the response from the consumer app e.g Compliance.

> It is way easier to add a custom function now rather than altering API responses

I think we do not have to think about the API at this point. What I am suggesting is to just put the processor you are suggesting into the consumer app itself rather than passing it through props hell :). I believe this is simpler approach. 

Anyways, I think this can be later addressed if there is an urge to implement the suggestion.

Reply:
Also, I feel we need to urge the API folks in general to have a little bit of consistency on their side as well. Otherwise, we will continue dealing with such API response discrepancies and writing processors on the frontend side. For example, OS response from the APIs are very different in shape if compared. Every app backend has its own shape :)